theano/sandbox/cuda/opt.py

```diff
@@ -2372,29 +2372,34 @@ def tensor_to_cuda(x):
 
 
 @register_opt()
-@local_optimizer([nlinalg.ExtractDiag])
+@local_optimizer([nlinalg.ExtractDiag, theano.tensor.Diagonal])
 def local_gpu_extract_diagonal(node):
     """
     extract_diagonal(host_from_gpu()) -> host_from_gpu(extract_diagonal)
 
     gpu_from_host(extract_diagonal) -> extract_diagonal(gpu_from_host)
 
     """
-    if (isinstance(node.op, nlinalg.ExtractDiag) and
-        isinstance(node.inputs[0].type,
-                   theano.tensor.TensorType)):
+    if ((isinstance(node.op, nlinalg.ExtractDiag) or
+         isinstance(node.op, theano.tensor.Diagonal)) and
+        isinstance(node.inputs[0].type, theano.tensor.TensorType)):
         inp = node.inputs[0]
+        if (isinstance(node.op, theano.tensor.Diagonal) and
+            (node.op.offset, node.op.axis1, node.op.axis2) != (0, 0, 1)):
+            warnings.warn("Diagonal Op does not non-default properties.")
```

Comment:
What about "Diagonal Op has no GPU implementation for non-default properties. It will run on CPU"